everestrs/everestrs/src/lib.rs

```diff
@@ -104,73 +123,147 @@ pub trait GenericModule: Sync {
+            .unwrap()
+            .handle_variable(impl_id, name, json.deserialize())
+            .unwrap();
+    }
+
+    pub fn publish_variable<T: serde::Serialize>(
+        &self,
+        impl_id: &str,
+        var_name: &str,
+        message: &T,
+    ) {
+        let blob = ffi::JsonBlob::from_vec(
+            serde_json::to_vec(&message).expect("Serialization of data cannot fail."),
+        );
+        (self.cpp_module)
+            .as_ref()
+            .unwrap()
+            .publish_variable(impl_id, var_name, blob);
+    }
+
+    pub fn call_command<T: serde::Serialize, R: serde::de::DeserializeOwned>(
+        &self,
+        impl_id: &str,
+        name: &str,
+        args: &T,
+    ) -> R {
+        let blob = ffi::JsonBlob::from_vec(
+            serde_json::to_vec(args).expect("Serialization of data cannot fail."),
+        );
+        let return_value = (self.cpp_module)
+            .as_ref()
+            .unwrap()
+            .call_command(impl_id, name, blob);
+        serde_json::from_slice(&return_value.data).unwrap()
+    }
+
     // TODO(hrapp): This function could use some error handling.
-    pub fn from_commandline<T: GenericModule + 'static>(module_impl: T) -> Self {
+    pub fn new() -> Self {
         let args: Args = argh::from_env();
-        let mut cpp_module = ffi::create_module(
+        let cpp_module = ffi::create_module(
             &args.module,
             &args.prefix.to_string_lossy(),
             &args.conf.to_string_lossy(),
         );
-        let manifest_json = cpp_module.as_mut().unwrap().initialize();
-        let manifest: schema::Manifest = manifest_json.deserialize();
-        let module = Self {
+
+        Self {
             cpp_module,
-            module_impl: Box::pin(module_impl),
-        };
+            sub_impl: RwLock::new(None),
+        }
+    }
+
+    pub fn set_subscriber(&self, sub_impl: Weak<dyn Subscriber>) {
+        *self.sub_impl.write().unwrap() = Some(sub_impl);
+        let manifest_json = self.cpp_module.as_ref().unwrap().initialize();
+        let manifest: schema::Manifest = manifest_json.deserialize();
 
         // Implement all commands for all of our implementations, dispatch everything to the
-        // GenericModule.
+        // Subscriber.
         for (implementation_id, implementation) in manifest.provides {
-            let interface_s = module.cpp_module.get_interface(&implementation.interface);
+            let interface_s = self.cpp_module.get_interface(&implementation.interface);
             let interface: schema::Interface = interface_s.deserialize();
             for (name, _) in interface.cmds {
-                let meta = ffi::CommandMeta {
-                    implementation_id: implementation_id.clone(),
-                    name,
-                };
+                self.cpp_module
+                    .as_ref()
+                    .unwrap()
+                    .provide_command(self, implementation_id.clone(), name);
```

Comment:
Question: The runtime (self) is passed here via `provide_command`.  Are rust references direct pointers, i.e. do they point to a specific memory location?  If so, this makes the Runtime class non-moveable, correct?

Reply:
This is a good point. But we have to pin not the subscriber reference but the runtime. The cxx bridge and set_subscribers should probably take Pin<Runtime> since this is the point from which we aren’t allowed to move. By not implementing Unpin we make sure that the runtime remains pinned once the set_subscriber is called